src/main.rs

```diff
@@ -539,30 +545,53 @@ fn notmain() -> Result<i32, anyhow::Error> {
 
     let debug_frame = debug_frame.ok_or_else(|| anyhow!("`.debug_frame` section not found"))?;
 
-    // print backtrace
-    let top_exception = backtrace(
-        &mut core,
-        pc,
-        debug_frame,
-        &elf,
-        &vector_table,
-        &sp_ram_region,
-        &live_functions,
-        &current_dir,
-    )?;
+    print_separator();
+
+    // check if the user enforced a backtrace print
+    let force_backtrace = match opts.force_backtrace.as_deref() {
+        Some("False") | Some("0") | Some("false") | None => false,
+        _ => true
+    };
+
+    let top_exception = construct_backtrace(
+            &mut core,
+            pc,
+            debug_frame,
+            &elf,
+            &vector_table,
+            &sp_ram_region,
+            &live_functions,
+            &current_dir,
+            // TODO any other cases in which we should force a backtrace?
+            force_backtrace || canary_touched
+        )?;
 
     core.reset_and_halt(TIMEOUT)?;
 
     Ok(
-        if let Some(TopException::HardFault { stack_overflow }) = top_exception {
-            if stack_overflow {
+        match top_exception {
+            Some(TopException::StackOverflow) => {
                 log::error!("the program has overflowed its stack");
+                SIGABRT
             }
+            Some(TopException::HardFault) => {
```

Comment:
Maybe name `TopException::Panic`, if it is indicating a panic?

Reply:
Hrrrm not sure– it's only indicating a panic because we know that `panic-probe` will trigger a HardFault on panic, no? I'd rather not bake this assumption further into our code– deducting "this was a panic" from "some HardFault occurred" is already muddying the waters imo so I'd like to contain this to this questionable match arm :D